src/chat.cpp

```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
+//
+bool chat_CanSendPrivateMessage( ULONG ulPlayer )
+{
+	if ( ulPlayer == MAXPLAYERS )
+		return true;
+
+	if ( g_ulChatPlayer == MAXPLAYERS )
+		g_ulChatPlayer = 0;
+
+	// [AK] If we're still trying to send a message to an invalid player, find another player.
+	if (( PLAYER_IsValidPlayer( g_ulChatPlayer ) == false ) || ( g_ulChatPlayer == (ULONG)consoleplayer ))
+	{
+		ULONG oldPlayer = g_ulChatPlayer;
+
+		do
+		{
+			// [AK] Keep looping until we find another player who we can send a message to.
+			if ( ++g_ulChatPlayer >= MAXPLAYERS )
+				g_ulChatPlayer = 0;
+
+			// [AK] If we're back to the old value, then there's nobody else to send a message to.
+			if ( g_ulChatPlayer == oldPlayer )
+			{
+				Printf( "There's no other player to send a private message to.\n" );
+				return false;
+			}
+		}
+		while (( PLAYER_IsValidPlayer( g_ulChatPlayer ) == false ) || ( g_ulChatPlayer == (ULONG)consoleplayer ));
+	}
+
+	return true;
+}
+
+void chat_PrivateMessage( FCommandLine &argv, const ULONG ulPlayer )
+{
+	ULONG		ulIdx;
+	FString		ChatString;
+
+	// [AK] Mods are not allowed to say anything in the player's name.
+	if ( ACS_IsCalledFromConsoleCommand( ) )
+		return;
+
+	// [AK] No chatting while playing a demo.
+	if ( CLIENTDEMO_IsPlaying( ) == true )
+	{
+		Printf( "You can't send private messages during demo playback.\n" );
+		return;
+	}
+
+	if ( argv.argc( ) >= 2 )
+	{
+		// [AK] Don't send private messages to invalid players.
+		if ( ulPlayer == MAXPLAYERS + 1 )
+		{
+			Printf( "There isn't a player named %s" TEXTCOLOR_NORMAL ".\n", argv[1] );
+			return;
+		}
+
+		// [AK] Don't send private messages to ourselves.
+		if (( ulPlayer == (ULONG)consoleplayer ) && ( NETWORK_GetState( ) != NETSTATE_SERVER ))
+		{
+			Printf( "You can't send a private message to yourself.\n" );
+			return;
+		}
+
+		g_ulChatPlayer = ulPlayer;
+
+		if ( argv.argc( ) > 2 )
+		{
+			for ( ulIdx = 2; ulIdx < static_cast<unsigned int>( argv.argc( ) ); ulIdx++ )
+				ChatString.AppendFormat( "%s ", argv[ulIdx] );
+
+			// Send the server's chat string out to clients, and print it in the console.
+			if ( NETWORK_GetState( ) == NETSTATE_SERVER )
+				SERVER_SendChatMessage( MAXPLAYERS, CHATMODE_PRIVATE_SEND, ChatString.GetChars( ), g_ulChatPlayer );
+			else
+				// We typed out our message in the console or with a macro. Go ahead and send the message now.
+				chat_SendMessage( CHATMODE_PRIVATE_SEND, ChatString.GetChars( ) );
+
+			return;
+		}
+	}
+	
+	if (( NETWORK_GetState( ) != NETSTATE_SERVER ) && ( chat_CanSendPrivateMessage( ulPlayer ) ))
+	{
+		// The message we send will only be for the player who we wish to chat with.
+		chat_SetChatMode( CHATMODE_PRIVATE_SEND );
+
+		// Hide the console.
+		C_HideConsole( );
+
+		// Clear out the chat buffer.
+		g_ChatBuffer.Clear( );
+	}
+}
+
+CCMD( sayto )
+{
+	ULONG ulPlayer = 0;
+
+	if ( argv.argc() >= 2 )
+	{
+		if ( stricmp( argv[1], "Server" ) == 0 )
```

Comment:
So if a player has the name "Server", this player will not be reachable by `sayto`, right? I don't think that this is really a problem, but one should just be aware of this.

Reply:
Technically yes, since it first assumes the client is sending a private message to the server rather than a player who may have the same name. I think it's unlikely somebody will try to use a name exactly like that (I remember some hosts would even forbid players from using it to prevent impersonation), but it's still possible. Should player names like "Server" or "<Server>" should even be allowed?